### ISSUE TITLE
NodeHealthy status is False, but machine is working properly (k0s)

### ISSUE THREAD

**Issue by nekwar** [kubernetes-sigs/cluster-api-provider-vsphere#2400]
/kind bug

**What steps did you take and what happened:**
Hi there
I'm trying to use CAPV in conjunction with k0smotron (https://docs.k0smotron.io/v0.6.0/cluster-api/). Definition of k0smotron control plane looks very similar to kubeadm:
```yaml
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
kind: K0smotronControlPlane
metadata:
  name: k0s-cluster
  namespace: default
spec:
  k0sVersion: v1.27.2-k0s.0
  persistence:
    type: emptyDir
  service:
    apiPort: 6443
    konnectivityPort: 8132
    type: LoadBalancer
```

So I've successfully deployed managed k0s cluster with one worker. Worker machine is provisioned (I can see that in status of `Machine` kind), managed cluster is operational. However, `NodeHealthy` status is 'false'. Here is the machine manifest (output of `kubectl get machine -o yaml`):

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Machine
metadata:
  creationTimestamp: '2023-09-27T12:12:00Z'
  finalizers:
    - machine.cluster.x-k8s.io
  generation: 3
  labels:
    cluster.x-k8s.io/cluster-name: k0s-cluster
    cluster.x-k8s.io/deployment-name: k0s-cluster-md-0
    cluster.x-k8s.io/set-name: k0s-cluster-md-0-7v9z5
    machine-template-hash: 4013674018-7v9z5
    node.cluster.x-k8s.io/esxi-host: vmw06-bud.ssl.mirantis.net
  managedFields:
    - apiVersion: cluster.x-k8s.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:finalizers:
            v:"machine.cluster.x-k8s.io": {}
          f:labels:
            f:cluster.x-k8s.io/cluster-name: {}
            f:cluster.x-k8s.io/deployment-name: {}
            f:cluster.x-k8s.io/set-name: {}
            f:machine-template-hash: {}
          f:ownerReferences:
            k:{"uid":"4e38463e-2ffc-4bc2-b3e9-ddeb15746f4b"}: {}
        f:spec:
          f:bootstrap:
            f:configRef: {}
          f:clusterName: {}
          f:infrastructureRef: {}
          f:version: {}
      manager: capi-machineset
      operation: Apply
      time: '2023-09-27T12:11:59Z'
    - apiVersion: cluster.x-k8s.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:labels:
            f:node.cluster.x-k8s.io/esxi-host: {}
        f:spec:
          f:bootstrap:
            f:dataSecretName: {}
          f:providerID: {}
      manager: manager
      operation: Update
      time: '2023-09-27T12:14:37Z'
    - apiVersion: cluster.x-k8s.io/v1beta1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          .: {}
          f:addresses: {}
          f:bootstrapReady: {}
          f:conditions: {}
          f:infrastructureReady: {}
          f:lastUpdated: {}
          f:observedGeneration: {}
          f:phase: {}
      manager: manager
      operation: Update
      subresource: status
      time: '2023-09-27T12:14:37Z'
  name: k0s-cluster-md-0-7v9z5-v5sww
  namespace: default
  ownerReferences:
    - apiVersion: cluster.x-k8s.io/v1beta1
      blockOwnerDeletion: true
      controller: true
      kind: MachineSet
      name: k0s-cluster-md-0-7v9z5
      uid: 4e38463e-2ffc-4bc2-b3e9-ddeb15746f4b
  resourceVersion: '3851426'
  uid: d2b60f00-837f-4ba2-bc3d-2b3ccb9f017c
  selfLink: >-
    /apis/cluster.x-k8s.io/v1beta1/namespaces/default/machines/k0s-cluster-md-0-7v9z5-v5sww
status:
  addresses:
    - address: 172.16.38.149
      type: ExternalIP
  bootstrapReady: true
  conditions:
    - lastTransitionTime: '2023-09-27T12:14:37Z'
      status: 'True'
      type: Ready
    - lastTransitionTime: '2023-09-27T12:12:22Z'
      status: 'True'
      type: BootstrapReady
    - lastTransitionTime: '2023-09-27T12:14:37Z'
      status: 'True'
      type: InfrastructureReady
    - lastTransitionTime: '2023-09-27T12:14:37Z'
      reason: NodeProvisioning
      severity: Warning
      status: 'False'
      type: NodeHealthy
  infrastructureReady: true
  lastUpdated: '2023-09-27T12:14:37Z'
  observedGeneration: 3
  phase: Provisioned
spec:
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
      kind: K0sWorkerConfig
      name: k0s-cluster-26sq8
      namespace: default
      uid: 43ff8928-64fa-4779-8e22-f65a7068d373
    dataSecretName: k0s-cluster-26sq8
  clusterName: k0s-cluster
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: VSphereMachine
    name: k0s-cluster-worker-jqw2d
    namespace: default
    uid: 5d72e8b1-e8d3-4a3f-aa68-0f3653aba5cf
  nodeDeletionTimeout: 10s
  providerID: vsphere://4215838e-5958-2c60-edcb-2dd9805db829
  version: v1.27.2+k0s.0
```

As you can see, there is one warning is status:
```yaml
    - lastTransitionTime: '2023-09-27T12:14:37Z'
      reason: NodeProvisioning
      severity: Warning
      status: 'False'
      type: NodeHealthy
```
My question is how this "health check" is performed and by which component?

**What did you expect to happen:**
Status of `NodeHealthy` to be `true` for MachineDeployment to properly managed this machine.

**Anything else you would like to add:**
There is nothing related to errors/warning in CAPI and CAPV controller managers logs. 

**Environment:**

- Cluster-api-provider-vsphere version: v1.5.1
- Kubernetes version: (use `kubectl version`): k0s kubelet --  v1.27.2+k0s.0
- OS (e.g. from `/etc/os-release`):  Ubuntu 20.04

**Comment by killianmuldoon** [comment]
`NodeHealthy` represents the health of the Kuberntes node that is running on the Machine. `NodeProvisioning` usually means that there's no Provider ID set on the node. Can you check if that is the case? 

The node Provider ID field is usually set by the external cloud provider. Are you running the VSphere Cloud Provider?

**Comment by nekwar** [comment]
@killianmuldoon thanks for the reply!

yes, you're right, `providerID` was not set on kubernetes node that is running on the Machine. 

> Are you running the VSphere Cloud Provider?

Yes, I have instance of Vsphere Cloud Controller Manager running, but for some reason, it's not setting `providerID` to the node. I took this value from Machine manifest (that I've posted earlier) and put it manually into the node definition, and now I can see `NodeHealthy` status `true` in management cluster. 

I will try to investigate further why is this happening. And also close this issue since further investigations are not related to CAPI/CAPV (it's more about Vsphere CCM). 

Thank you for your help!
